### ISSUE TITLE
WWW-Authenticate not returned to client

### ISSUE THREAD

**Issue by ptaferner** [civetweb/civetweb#734]
For a business project I have the requirement to do an AJAX call against a digest auth protected lua script. Handling digest authentication in JS wouldn't be the problem, as it is possible by parsing the nonce from the first unauthenticated response.
However, the necessary WWW-Authenticate header is not returned to the client from my Civetweb application, although I can cleary see it in a packet sniffer like Wireshark. By "not returned to the client" I mean that I cannot get it via getAllResponseHeaders from the XMLHttpRequest object.

A different approach to this problem would be to use basic authentication - then I wouldn't need the not so trivial handling of digest authentication via Javascript. However, it seems that Civetweb currently only supports digest authentication, right?
There used to be issue #253 which mentioned the need for basic authentication, but it's a few years old.
I could also do some token based authentication for this AJAX request, but this seems also a rather complicated solution.

Note: I'm not usign CORS here which would possibly explain the missing headers.

**Comment by bel2125** [comment]
Basic auth:
In my experience, most *browser applications* use cookie based auth, while *M2M applications* tend to use certificate based authentication. A small number of applications uses digest. Some browsers do not even allow to use basic auth for HTTP but only for HTTPS, since the password is basically clear text. So, there is currently no support for basic auth in civetweb and it's currently not on my roadmap. I might be wrong with my assessment "nobody need basic auth", and will be happy if someone explains my why.

If you are building the server yourself (instead of using the pre-compiled executable from sourceforge), you can use the C interface `mg_set_auth_handler` to register a C handler callback function. The handler function is called every time a request is made to the given URI pattern. In this handler you can decide to either accept the request (based on the request header), deny it, ask for authentication (digest or even basic), redirect it to some login page to obtain a cookie, ... check your accounting database and allow access only if the membership fee is payed, or the time of access corresponds to the payed plan ... whatever someone can imagine (endless possibilities to torment the users).

**Comment by ptaferner** [comment]
Thanks so far.
The authentication used in this project is kind of legacy, I know - I suggested cookies already to the product management, but so far it hasn't been considered.
And even if I would use cookies for that - currently my web application supports accepting a PUT request for uploading a larger file. According to the documentation, PUT requests need to be subject to digest authentication via a configured password file. So at least for that I need HTTP authentication, it seems - unless I consider configuring (patching) the PUT handler for that.

This PUT request is actually the root of all evil here concerning my problem, specifically the combination of a PUT request with HTTPS and digest authentication: For some reason Civetweb returns a TCP reset directly after the 401 (which would be the proper response to the first unauthenticated PUT request of the client).

To sum up: If I switched to cookie-based authentication which seems feasible to me in my situation, then I would have to patch the PUT handler. But this could still be the least time-consuming way...

**Comment by bel2125** [comment]
If you use a Lua handler, callback, CGI script sub-path ... basically everything that qualifies as "script resource", your PUT request will be forwarded to this script. just like a POST request.
If you send a PUT to CivetWeb, it will internally try to distinguish between two situations:

a) You want to store (add or replace) a file in the document root directory of the server. This is a (not recommended) way of server administration for simple servers, and you need to authenticate as the administrator of the server.

b) The server presents a virtual object, say a collection of resources exposed as URI, but not corresponding to a file. The resource collection is managed by a callback or script - any request that you send to an URI that is a subdirectory of the resource management script will be forwarded to this script ... GET, POST, also PUT, DELETE, PATCH, ... all HTTP methods (verbs) that are *in principle* valid.

So, if you are sending a PUT request to something that is a sub-path of a script, you can handle the PUT in whatever way you like - you can ask for basic auth there, if you like.

**Comment by ptaferner** [comment]
I've handled the problem by introducing cookie-based authentication for browser communication, and the digest authentication only for API access. This has been planned for the future anyway; I just did it a bit earlier now.
Thanks for your valuable input so far - this issue can be closed now, I suppose.
